### PR TITLE
Make torch optional. Update DragonNet w/ latest TF APIs

### LINKED ISSUE #789: Make torch an optional dependency
**Is your feature request related to a problem? Please describe.**

The `torch` package, which is currently a required dependency for `causalml`, requires presence of CUDA libraries like `libcublas.so.*[0-9]`. This causes installation issues, especially in environments where these libraries are not pre-installed or compatible, leading to errors such as `ValueError: libcublas.so.*[0-9] not found in the system path`. However, within `causalml`, `torch` is [only used for the Causal Effect VAE model](https://github.com/search?q=repo%3Auber%2Fcausalml%20torch&type=code), which relies on neural networks. 

`torch` is also a large package, would be good to not have to install it by default.

**Describe the solution you'd like**
To resolve this, I propose making `torch` an optional dependency within `causalml` via `pip install causalml[torch]`. This would be similar to the previously implemented approach with TensorFlow in https://github.com/uber/causalml/pull/343, where it was made optional and only installed when needed via `pip install causalml[tf]`. 

**Describe alternatives you've considered**
* A workaround for envs with no GPUs is to specify the following in your Pipfile when you want to add `causalml`: 
```
[[source]]
name = "pytorch"
url = "https://download.pytorch.org/whl/cpu"
verify_ssl = true

torch = {version = "==2.1.2", index = "pytorch"}
```
* A better, but way more involved, alternative would be to separate training deps from prediction deps as suggested in: https://github.com/uber/causalml/issues/594

### PR DESCRIPTION
**Pull request by jeongyoonlee**
## Proposed changes

This PR fixes #789 by making the torch dependency optional.

It also fixes the `DragonNet` test error with the latest Tensorflow APIs.
 
## Types of changes

What types of changes does your code introduce to **CausalML**?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/uber/causalml/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/uber/causalml)
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

## Further comments

N/A

### CODE REVIEW

File: tests/test_dragonnet.py [hunk 1]
```diff
@@ -3,21 +3,21 @@
...
-def test_save_load_dragonnet():
+def test_save_load_dragonnet(tmp_path):
     y, X, w, tau, b, e = simulate_nuisance_and_easy_treatment(n=1000)
 
     dragon = DragonNet(neurons_per_layer=200, targeted_reg=True, verbose=False)
     dragon_ite = dragon.fit_predict(X, w, y, return_components=False)
     dragon_ate = dragon_ite.mean()
-    dragon.save("smaug")
+
+    model_file = tmp_path / "smaug.h5"
```
Comment (paullo0106) [inline tests/test_dragonnet.py:17]: @jeongyoonlee did you mean `f"{tmp_path}/smaug.h5"` ? I'm not super familiar with the code here

Comment (jeongyoonlee) [reply tests/test_dragonnet.py:17]: tmp_path is a Pathlib object and this is a valid syntax. :)

**Review by paullo0106 — APPROVED** [review]
Thanks!

File: tests/test_dragonnet.py (hunk 1 above)
Comment (paullo0106) [reply tests/test_dragonnet.py:17]: i see, makes sense then!

**Pull request merged**